source-code/core2/src/messages/query.ts

```diff
@@ -1,15 +1,54 @@
+import { createInlang } from "../app/createInlang.js"
+import type { LanguageTag } from "../languageTag.js"
 import type { Message } from "./schema.js"
 
-// TODO
-// 1. Query builder?
-// 2. Filtering etc how?
-//
+type UniqueFilterType = {
+	id: Message["id"]
+	languageTag: LanguageTag
+	//variant: string -> in the future
+}
+type FilterType = {
+	id?: Message["id"]
+	languageTag?: LanguageTag
+	//variant?: string -> in the future
+}
+
 // inspiration: Prisma CRUD https://www.prisma.io/docs/concepts/components/prisma-client/crud
 export type MessagesQueryApi_1 = {
-	create: (message: Message) => void
-	get: () => Message[]
-	update: () => void
-	delete: () => void
+	get: (args: { where: UniqueFilterType }) => [Message, Error]
+	getMany: (args: { where: FilterType }) => [Message[], Error]
+	upsert: (args: Message) => [Message, Error]
+	upsertMany: (args: Message[]) => [Message[], Error]
+	delete: (args: { where: UniqueFilterType }) => [Message, Error]
+	deleteMany: (args: { where: FilterType }) => [Message[], Error]
 }
 
-// query.
+const inlang = createInlang({
+	configPath: "/hello/inlang.config.json",
+	env: { fs: undefined as any, import: undefined as any },
+})
+
+// create
+const message1 = inlang.messages.query.upsert({
+	id: "myMessageId",
+	languageTag: "en",
+	pattern: [{ type: "Text", value: "Hello World" }],
+})
+
+const message2 = inlang.messages.query.get({ where: { id: "myMessageId", languageTag: "en" } })
```

Comment:
can we get rid of `.query`?
`inlang.messages.get(/* ... */)`

Reply:
We discussed having a query2 in the future, which can be easily added to this setup to introduce reactivity or caching.

Reply:
I vote for `inlang.messages.query.*`

The proposal would add 10 functions to the messages object. If we decide to add query unrelated props, the auto complete experience will be degraded because query functions mix with other properties. 

FYI we can easily move `messages.query` props to `messages` if its so painful. We can't easily move `messages` functions to `messages.query` without having an endless list of deprecated props on autocompletion.

Reply:
Fair enough 👍

Reply:
After https://github.com/inlang/inlang/issues/1159, I am 100% convinces `messages.query.*` is the way to go to avoid breaking changes. Changing the query API because new requirements emerge can break the entire ecosystem.